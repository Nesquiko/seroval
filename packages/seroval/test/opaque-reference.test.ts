import { describe, expect, it } from 'vitest';
import {
  OpaqueReference,
  crossSerialize,
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromCrossJSON,
  fromJSON,
  serialize,
  serializeAsync,
  toCrossJSON,
  toCrossJSONAsync,
  toCrossJSONStream,
  toJSON,
} from '../src';

const EXAMPLE = {
  transparent: 'This is transparent',
  opaque: new OpaqueReference('This is opaque'),
};

const REPLACEMENT = 'This is a dummy value.';

const EXAMPLE_WITH_REPLACEMENT = {
  transparent: 'This is transparent',
  opaque: new OpaqueReference('This is opaque', REPLACEMENT),
};

describe('OpaqueReference', () => {
  describe('serialize', () => {
    it('supports OpaqueReference', () => {
      const result = serialize(EXAMPLE);
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof EXAMPLE>(result);
      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE.transparent);
      expect(back.opaque).toBe(undefined);
    });
    it('supports OpaqueReference with replacement', () => {
      const result = serialize(EXAMPLE_WITH_REPLACEMENT);
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof EXAMPLE_WITH_REPLACEMENT>(result);
      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE_WITH_REPLACEMENT.transparent);
      expect(back.opaque).toBe(REPLACEMENT);
    });
  });
  describe('serializeAsync', () => {
    it('supports OpaqueReference', async () => {
      const result = await serializeAsync(EXAMPLE);
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof EXAMPLE>(result);
      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE.transparent);
      expect(back.opaque).toBe(undefined);
    });
    it('supports OpaqueReference with replacement', async () => {
      const result = await serializeAsync(EXAMPLE_WITH_REPLACEMENT);
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof EXAMPLE_WITH_REPLACEMENT>(result);
      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE_WITH_REPLACEMENT.transparent);
      expect(back.opaque).toBe(REPLACEMENT);
    });
  });
  describe('toJSON', () => {
    it('supports OpaqueReference', () => {
      const result = toJSON(EXAMPLE);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof EXAMPLE>(result);
      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE.transparent);
      expect(back.opaque).toBe(undefined);
    });
    it('supports OpaqueReference with replacement', () => {
      const result = toJSON(EXAMPLE_WITH_REPLACEMENT);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof EXAMPLE_WITH_REPLACEMENT>(result);
      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE_WITH_REPLACEMENT.transparent);
      expect(back.opaque).toBe(REPLACEMENT);
    });
  });
  describe('toJSONAsync', () => {
    it('supports OpaqueReference', () => {
      const result = toJSON(EXAMPLE);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof EXAMPLE>(result);
      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE.transparent);
      expect(back.opaque).toBe(undefined);
    });
    it('supports OpaqueReference with replacement', () => {
      const result = toJSON(EXAMPLE_WITH_REPLACEMENT);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof EXAMPLE_WITH_REPLACEMENT>(result);
      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE_WITH_REPLACEMENT.transparent);
      expect(back.opaque).toBe(REPLACEMENT);
    });
  });
  describe('crossSerialize', () => {
    it('supports OpaqueReference', () => {
      const result = crossSerialize(EXAMPLE);
      expect(result).toMatchSnapshot();
    });
    it('supports OpaqueReference with replacement', () => {
      const result = crossSerialize(EXAMPLE_WITH_REPLACEMENT);
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports OpaqueReference', () => {
        const result = crossSerialize(EXAMPLE, { scopeId: 'example' });
        expect(result).toMatchSnapshot();
      });
      it('supports OpaqueReference with replacement', () => {
        const result = crossSerialize(EXAMPLE_WITH_REPLACEMENT);
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports OpaqueReference', async () => {
      const result = await crossSerializeAsync(EXAMPLE);
      expect(result).toMatchSnapshot();
    });
    it('supports OpaqueReference with replacement', async () => {
      const result = await crossSerializeAsync(EXAMPLE_WITH_REPLACEMENT);
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports OpaqueReference', async () => {
        const result = await crossSerializeAsync(EXAMPLE, {
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
      it('supports OpaqueReference with replacement', async () => {
        const result = await crossSerializeAsync(EXAMPLE_WITH_REPLACEMENT, {
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports OpaqueReference', async () =>
      new Promise<void>((resolve, reject) => {
        crossSerializeStream(EXAMPLE, {
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    it('supports OpaqueReference with replacement', async () =>
      new Promise<void>((resolve, reject) => {
        crossSerializeStream(EXAMPLE_WITH_REPLACEMENT, {
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    describe('scoped', () => {
      it('supports OpaqueReference', async () =>
        new Promise<void>((resolve, reject) => {
          crossSerializeStream(EXAMPLE, {
            scopeId: 'example',
            onSerialize(data) {
              expect(data).toMatchSnapshot();
            },
            onDone() {
              resolve();
            },
            onError(error) {
              reject(error);
            },
          });
        }));
      it('supports OpaqueReference with replacement', async () =>
        new Promise<void>((resolve, reject) => {
          crossSerializeStream(EXAMPLE_WITH_REPLACEMENT, {
            scopeId: 'example',
            onSerialize(data) {
              expect(data).toMatchSnapshot();
            },
            onDone() {
              resolve();
            },
            onError(error) {
              reject(error);
            },
          });
        }));
    });
  });
  describe('toCrossJSON', () => {
    it('supports OpaqueReference', () => {
      const result = toCrossJSON(EXAMPLE);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof EXAMPLE>(result, {
        refs: new Map(),
      });

      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE.transparent);
      expect(back.opaque).toBe(undefined);
    });
    it('supports OpaqueReference with replacement', () => {
      const result = toCrossJSON(EXAMPLE_WITH_REPLACEMENT);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof EXAMPLE_WITH_REPLACEMENT>(result, {
        refs: new Map(),
      });

      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE_WITH_REPLACEMENT.transparent);
      expect(back.opaque).toBe(REPLACEMENT);
    });
  });
  describe('toCrossJSONAsync', () => {
    it('supports OpaqueReference', async () => {
      const result = await toCrossJSONAsync(EXAMPLE);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof EXAMPLE>(result, {
        refs: new Map(),
      });

      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE.transparent);
      expect(back.opaque).toBe(undefined);
    });
    it('supports OpaqueReference with replacement', async () => {
      const result = await toCrossJSONAsync(EXAMPLE_WITH_REPLACEMENT);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof EXAMPLE_WITH_REPLACEMENT>(result, {
        refs: new Map(),
      });

      expect(back.constructor).toBe(Object);
      expect(back.transparent).toBe(EXAMPLE_WITH_REPLACEMENT.transparent);
      expect(back.opaque).toBe(REPLACEMENT);
    });
  });
  describe('toCrossJSONStream', () => {
    it('supports OpaqueReference', async () =>
      new Promise<void>((resolve, reject) => {
        toCrossJSONStream(EXAMPLE, {
          onParse(data) {
            expect(JSON.stringify(data)).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    it('supports OpaqueReference with replacement', async () =>
      new Promise<void>((resolve, reject) => {
        toCrossJSONStream(EXAMPLE_WITH_REPLACEMENT, {
          onParse(data) {
            expect(JSON.stringify(data)).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
  });
});
