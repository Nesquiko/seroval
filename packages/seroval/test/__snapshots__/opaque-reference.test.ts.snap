// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OpaqueReference > crossSerialize > scoped > supports OpaqueReference 1`] = `"($R=>$R[0]={transparent:"This is transparent",opaque:void 0})($R["example"])"`;

exports[`OpaqueReference > crossSerialize > scoped > supports OpaqueReference with replacement 1`] = `"($R[0]={transparent:"This is transparent",opaque:"This is a dummy value."})"`;

exports[`OpaqueReference > crossSerialize > supports OpaqueReference 1`] = `"($R[0]={transparent:"This is transparent",opaque:void 0})"`;

exports[`OpaqueReference > crossSerialize > supports OpaqueReference with replacement 1`] = `"($R[0]={transparent:"This is transparent",opaque:"This is a dummy value."})"`;

exports[`OpaqueReference > crossSerializeAsync > scoped > supports OpaqueReference 1`] = `"($R=>$R[0]={transparent:"This is transparent",opaque:void 0})($R["example"])"`;

exports[`OpaqueReference > crossSerializeAsync > scoped > supports OpaqueReference with replacement 1`] = `"($R=>$R[0]={transparent:"This is transparent",opaque:"This is a dummy value."})($R["example"])"`;

exports[`OpaqueReference > crossSerializeAsync > supports OpaqueReference 1`] = `"($R[0]={transparent:"This is transparent",opaque:void 0})"`;

exports[`OpaqueReference > crossSerializeAsync > supports OpaqueReference with replacement 1`] = `"($R[0]={transparent:"This is transparent",opaque:"This is a dummy value."})"`;

exports[`OpaqueReference > crossSerializeStream > scoped > supports OpaqueReference 1`] = `"($R=>$R[0]={transparent:"This is transparent",opaque:void 0})($R["example"])"`;

exports[`OpaqueReference > crossSerializeStream > scoped > supports OpaqueReference with replacement 1`] = `"($R=>$R[0]={transparent:"This is transparent",opaque:"This is a dummy value."})($R["example"])"`;

exports[`OpaqueReference > crossSerializeStream > supports OpaqueReference 1`] = `"($R[0]={transparent:"This is transparent",opaque:void 0})"`;

exports[`OpaqueReference > crossSerializeStream > supports OpaqueReference with replacement 1`] = `"($R[0]={transparent:"This is transparent",opaque:"This is a dummy value."})"`;

exports[`OpaqueReference > serialize > supports OpaqueReference 1`] = `"({transparent:"This is transparent",opaque:void 0})"`;

exports[`OpaqueReference > serialize > supports OpaqueReference with replacement 1`] = `"({transparent:"This is transparent",opaque:"This is a dummy value."})"`;

exports[`OpaqueReference > serializeAsync > supports OpaqueReference 1`] = `"({transparent:"This is transparent",opaque:void 0})"`;

exports[`OpaqueReference > serializeAsync > supports OpaqueReference with replacement 1`] = `"({transparent:"This is transparent",opaque:"This is a dummy value."})"`;

exports[`OpaqueReference > toCrossJSON > supports OpaqueReference 1`] = `"{"t":10,"i":0,"p":{"k":["transparent","opaque"],"v":[{"t":1,"s":"This is transparent"},{"t":2,"s":1}],"s":2},"o":0}"`;

exports[`OpaqueReference > toCrossJSON > supports OpaqueReference with replacement 1`] = `"{"t":10,"i":0,"p":{"k":["transparent","opaque"],"v":[{"t":1,"s":"This is transparent"},{"t":1,"s":"This is a dummy value."}],"s":2},"o":0}"`;

exports[`OpaqueReference > toCrossJSONAsync > supports OpaqueReference 1`] = `"{"t":10,"i":0,"p":{"k":["transparent","opaque"],"v":[{"t":1,"s":"This is transparent"},{"t":2,"s":1}],"s":2},"o":0}"`;

exports[`OpaqueReference > toCrossJSONAsync > supports OpaqueReference with replacement 1`] = `"{"t":10,"i":0,"p":{"k":["transparent","opaque"],"v":[{"t":1,"s":"This is transparent"},{"t":1,"s":"This is a dummy value."}],"s":2},"o":0}"`;

exports[`OpaqueReference > toCrossJSONStream > supports OpaqueReference 1`] = `"{"t":10,"i":0,"p":{"k":["transparent","opaque"],"v":[{"t":1,"s":"This is transparent"},{"t":2,"s":1}],"s":2},"o":0}"`;

exports[`OpaqueReference > toCrossJSONStream > supports OpaqueReference with replacement 1`] = `"{"t":10,"i":0,"p":{"k":["transparent","opaque"],"v":[{"t":1,"s":"This is transparent"},{"t":1,"s":"This is a dummy value."}],"s":2},"o":0}"`;

exports[`OpaqueReference > toJSON > supports OpaqueReference 1`] = `"{"t":{"t":10,"i":0,"p":{"k":["transparent","opaque"],"v":[{"t":1,"s":"This is transparent"},{"t":2,"s":1}],"s":2},"o":0},"f":31,"m":[]}"`;

exports[`OpaqueReference > toJSON > supports OpaqueReference with replacement 1`] = `"{"t":{"t":10,"i":0,"p":{"k":["transparent","opaque"],"v":[{"t":1,"s":"This is transparent"},{"t":1,"s":"This is a dummy value."}],"s":2},"o":0},"f":31,"m":[]}"`;

exports[`OpaqueReference > toJSONAsync > supports OpaqueReference 1`] = `"{"t":{"t":10,"i":0,"p":{"k":["transparent","opaque"],"v":[{"t":1,"s":"This is transparent"},{"t":2,"s":1}],"s":2},"o":0},"f":31,"m":[]}"`;

exports[`OpaqueReference > toJSONAsync > supports OpaqueReference with replacement 1`] = `"{"t":{"t":10,"i":0,"p":{"k":["transparent","opaque"],"v":[{"t":1,"s":"This is transparent"},{"t":1,"s":"This is a dummy value."}],"s":2},"o":0},"f":31,"m":[]}"`;
